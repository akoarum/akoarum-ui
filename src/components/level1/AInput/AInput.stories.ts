import { storiesOf } from '@storybook/vue'
import { action } from '@storybook/addon-actions'
import { withKnobs, text } from '@storybook/addon-knobs'
import { withInfo } from 'storybook-addon-vue-info'
import AInput from './AInput.vue'

storiesOf('level1/AInput', module)
  .addDecorator(withInfo)
  .addDecorator(withKnobs)
  .add('Normal', () => ({
    components: { AInput },
    props: {
      type: { default: text('Type', 'text') },
      width: { default: text('Width', 'auto') }
    },
    data() {
      return {
        value: ''
      }
    },
    template: `
      <AInput v-model="value" :type="type" :width="width" @focus="focus" @blur="blur" />
    `,
    methods: {
      focus: action('focus'),
      blur: action('blur')
    }
  }), { info: {} })
  .add('Error', () => ({
    components: { AInput },
    props: {
      type: { default: text('Type', 'text') },
      width: { default: text('Width', 'auto') }
    },
    data() {
      return {
        value: ''
      }
    },
    template: `
      <AInput v-model="value" :type="type" :width="width" aria-invalid="true" @focus="focus" @blur="blur" />
    `,
    methods: {
      focus: action('focus'),
      blur: action('blur')
    }
  }), { info: {} })
