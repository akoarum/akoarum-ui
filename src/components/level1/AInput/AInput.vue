<template>
  <input
    v-model="modelValue"
    :style="styled"
    class="aInput"
    @focus="handleFocus"
    @blur="handleBlur"
  />
</template>

<script lang="ts">
import { defineComponent, computed, SetupContext } from '@vue/composition-api'

interface Props {
  value: string
  width?: string
  error?: boolean
}

export default defineComponent({
  props: {
    value: { type: String, required: true },
    width: { type: String }
  },
  setup(props: Props, context: SetupContext) {
    const styled = computed<{ [key: string]: string }>(() => {
      const styles: { [key: string]: string } = {}
      if (props.width) styles['--width'] = props.width
      return styles
    })
    const modelValue = computed({
      get: (): string => props.value,
      set: (newValue: string) => {
        context.emit('input', newValue)
      }
    })
    const handleFocus = () => {
      context.emit('focus')
    }
    const handleBlur = () => {
      context.emit('blur')
    }

    return { styled, modelValue, handleFocus, handleBlur }
  }
})
</script>

<style lang="scss" scoped>
.aInput {
  --width: auto;
  display: inline-block;
  box-sizing: border-box;
  border: 1px solid $color-border;
  box-shadow: none;
  padding: 8px 12px 9px;
  width: var(--width);
  font-size: 16px;
  line-height: 1.3;
  outline: none;
  transition: border $duration $easing, background $duration $easing;

  &[aria-invalid='true'] {
    border-color: $color-caution;
    background: #fff7f7;
  }

  &:focus {
    border-color: $color-primary;
  }
}
</style>
