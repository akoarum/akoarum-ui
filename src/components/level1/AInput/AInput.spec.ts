import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueCompositionApi from '@vue/composition-api'
import AInput from './AInput.vue'

describe('AInput', () => {
  const localVue = createLocalVue()
  let wrapper

  localVue.use(VueCompositionApi)

  beforeEach(() => {
    wrapper = shallowMount(AInput, {
      propsData: {
        value: ''
      },
      localVue
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })

    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    it('styled: propsで幅をセットできる', () => {
      wrapper.setProps({ width: '100%' })
      expect(wrapper.vm.styled).toEqual({ '--width': '100%' })
    })
  })

  describe('event', () => {
    it('handleFocus: フォーカスイベントを通知する', () => {
      wrapper.vm.handleFocus()
      expect(wrapper.emitted().focus).toBeTruthy()
    })

    it('handleBlur: フォーカスが外れたことを通知する', () => {
      wrapper.vm.handleBlur()
      expect(wrapper.emitted().blur).toBeTruthy()
    })
  })
})
