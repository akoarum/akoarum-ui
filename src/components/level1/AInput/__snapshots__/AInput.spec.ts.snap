// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AInput initialized snapshot 1`] = `
<input
  class="aInput"
/>
`;
